app/src/main/java/org/greenstand/android/TreeTracker/managers/Locations.kt

```diff
@@ -0,0 +1,332 @@
+    }
+}
+
+enum class Accuracy {
+    GOOD,
+    BAD,
+    NONE
+}
+
+fun Location?.accuracyStatus(): Accuracy {
+    if (this == null || !hasAccuracy()) {
+        return Accuracy.NONE
+    }
+    return if (accuracy < ValueHelper.MIN_ACCURACY_DEFAULT_SETTING) {
+        Accuracy.GOOD
+    } else {
+        Accuracy.BAD
+    }
+}
+
+class LocationDataCapturer(
+    private val userManager: UserManager,
+    private val locationUpdateManager: LocationUpdateManager,
+    private val treeTrackerDAO: TreeTrackerDAO
+) {
+    var convergenceWithinRange: Boolean = false
+
+    private val gson = GsonBuilder().serializeNulls().create()
+    private var locationsDeque: Deque<Location> = LinkedList<Location>()
+    var generatedTreeUuid: UUID? = null
+        private set
+    var convergence: Convergence? = null
+
+    private val locationObserver: Observer<Location?> = Observer {
+        it?.apply {
+
+            MainScope().launch(Dispatchers.IO) {
+                val locationData =
+                    LocationData(
+                        userManager.planterCheckinId,
+                        latitude,
+                        longitude,
+                        accuracy,
+                        generatedTreeUuid?.toString() ?: null,
+                        System.currentTimeMillis()
+                    )
+                val jsonValue = gson.toJson(locationData)
+                Timber.d("Inserting new location data $jsonValue")
+                treeTrackerDAO.insertLocationData(LocationDataEntity(jsonValue))
+            }
+
+            if (isInTreeCaptureMode() && !convergenceWithinRange) {
+
+                var evictedLocation: Location? = if (locationsDeque.size >= CONVERGENCE_DATA_SIZE)
+                    locationsDeque.pollFirst() else null
+                locationsDeque.add(it)
+
+                if (locationsDeque.size >= CONVERGENCE_DATA_SIZE) {
+                    if (convergence == null) {
+                        convergence = Convergence(locationsDeque.toList())
+                        convergence!!.computeConvergence()
+                    } else {
+                        convergence!!.computeSlidingWindowConvergence(evictedLocation!!, it)
+                    }
+                    Timber.d(
+                        "Convergence: Longitude Mean: " +
+                                "[${convergence!!.longitudeConvergence?.mean}]. \n" +
+                                "Longitude standard deviation value: " +
+                                "[${convergence!!.longitudeConvergence?.standardDeviation}]"
+                    )
+                    Timber.d(
+                        "Convergence: Latitude Mean: " +
+                                "[${convergence!!.latitudeConvergence?.mean}]. \n " +
+                                "Latitude standard deviation value: " +
+                                "[${convergence!!.latitudeConvergence?.standardDeviation}]"
+                    )
+                    safeLet(
```

Comment:
Can we remove this safeLet? It's only used in one place.

`
val lat = convergence?.longitudinalStandardDeviation() ?: return

val long = convergence?.latitudinalStandardDeviation() ?: return

convergenceWithinRange = long < LONG_STD_DEV && lat < LAT_STD_DEV
`

Reply:
Updated this and removed `safeLet`